Include button elements in DOM XSS Scan Rule

mrod23 opened the pull request:
The goal of this MR is to have the DomXSS Plugin click on additional buttons within the DOM to see if XSS can be triggered.

The script is hitting buttons like this: `<button onclick="onCancelClicked()">Click me</button>` , but doesn't test buttons formatted like this: but not like this `<button type="button">Cancel</button>`

Tested against an internal site, screenshot here: 
![image](https://user-images.githubusercontent.com/3794761/100753797-41733380-33b8-11eb-9754-f91dcfa1f1c2.png)

mrod23 commented:
The goal of this MR is to have the DomXSS Plugin click on additional buttons within the DOM to see if XSS can be triggered. 

The script is hitting buttons like this: 
<button onclick="onCancelClicked()">Click me</button>

but not like this 
<button type="button">Cancel</button>

psiinon commented on addOns/domxss/src/main/java/org/zaproxy/zap/extension/domxss/DomXssScanRule.java:
```diff
@@ -453,9 +453,12 @@ private void scanHelper(String attackVector, String url) throws DomAlertExceptio
         try {
             inputElements = findHelper(driver, By.tagName("input"));
+            buttonElements = findHelper(driver, By.tagName("button"));
+            inputElements.addAll(buttonElements);
```
You can combine this into just one line:
`inputElements.addAll(findHelper(driver, By.tagName("button")))`

mrod23 replied:
Will do. I'll combine

mrod23 replied:
combined

psiinon commented:
I like the principal here - theres definiately more chance of hitting DOM XSSs if we click on the buttons.
Can we tell they are buttons and not sendKeys to them? 
Strictly speaking they will no longer all be input elements so maybe rename that var and if possible test that they are input elements before calling sendKeys(..).
Can you explain how it avoiding to click on cancel buttons?

thc202 commented on addOns/domxss/CHANGELOG.md:
```diff
@@ -15,6 +15,10 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
+## [10] - 2020-12-01
+### Changed
+- Now clicking on different buttons throughout the page to see if it triggers XSS.
```
This should be added to the above "Changed" section.

mrod23 replied:
> Can you explain how it avoiding to click on cancel buttons?
> Can you explain how it avoiding to click on cancel buttons?

The main reason I wanted to add this is so the plugin will click the cancel buttons. We have JavaScript similar to the below on one of our sites. When the Cancel button is clicked, the JavaScript that's in the retUrl attribute is triggered on the user's page. 
```
            $("#DIAGLOG").dialog("option", "buttons", {
                "Cancel": function () {
                    window.location = '@retUrl';
                },
                "Submit": function () {
                    xxxxxxxx
                    });
                }
            });
```

mrod23 replied:
> Can we tell they are buttons and not sendKeys to them?
What's an example of a sendKey?

thc202 commented:
Worth adding tests.

mrod23 commented:
@thc202 
> Worth adding tests.

What kind of tests are you thinking?

thc202 commented:
Tests that verify the buttons are actually clicked, with a case where it leads to XSS and another that doesn't.

mrod23 commented:
@psiinon
>    Can you explain how it avoiding to click on cancel buttons?

The main reason I wanted to add this is so the plugin will click the cancel buttons. We have JavaScript similar to the below on one of our sites. When the Cancel button is clicked, the JavaScript that's in the retUrl attribute is triggered on the user's page.

```
            $("#DIAGLOG").dialog("option", "buttons", {
                "Cancel": function () {
                    window.location = '@retUrl';
                },
                "Submit": function () {
                    xxxxxxxx
                    });
                }
            });
```

>    Can we tell they are buttons and not sendKeys to them?

What's an example of a sendKey?

psiinon commented:
https://github.com/zaproxy/zap-extensions/blob/41d43cb9eb93bff3a4fe2dd58c304f49dbb2b779/addOns/domxss/src/main/java/org/zaproxy/zap/extension/domxss/DomXssScanRule.java#L482 :)
sendKeys( ) is the method that submits the input to input fields, which obviously doesnt apply to buttons.
And +1 for tests - ones which find vulns. We have a reasonable set of tests for this add-on but just shout if you need some help.

mrod23 commented:
Ah the sendKeys() method. Sounds good I can put in a check . 

I may need help with the unit tests. Not seeing a test page I can use here: **http://public-firing-range.appspot.com**

psiinon commented:
No, I imported all of the DOM XSS vulns that still worked from that site :D
They are all here: https://github.com/zaproxy/zap-extensions/tree/master/addOns/domxss/src/test/resources/org/zaproxy/zap/extension/domxss
If you can create some single pages like those which demo the vulns which require button input then we can help you create tests for those.

mrod23 commented:
@psiinon 

I created a test page and added to the Resources directory. Zap will only find the XSS with the changes added in this MR. The current version shouldn't be able to find them. 

 In order for Zap to trigger an alert on the test page it requires a URL parameter called "returnUrl". 
Something like this: https://localhost:8080/test/jquery-3.html?returnUrl=test

I couldn't figure out how to get the Unit test to query a page with a URL parameter. Can you help me out with that? I started a unit test, but I have it commented out. 

I addressed the other comments mentioned above in the latest PR.

kingthorin commented:
@mrod23 have a look at https://github.com/zaproxy/zap-extensions/blob/master/addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRuleUnitTest.java it should have details to help you out.

mrod23 commented:
Thanks @kingthorin 

@kingthorin  / @psiinon  can you take a look at my branch when you get a chance?
The unit test is still failing, pasted what the failure output looks like 

I think I need to update when the code raises an assertAlert here: https://github.com/zaproxy/zap-extensions/blob/ff51107d3abf016a51845b065ac3523abe9a1e5c/addOns/domxss/src/test/java/org/zaproxy/zap/extension/domxss/DomXssScanRuleUnitTest.java#L127

Tried a few different combinations of new variables but I couldn't get it to trigger. 

If you serve up the CancelButton page locally, and run Zap with the plugin built from this branch it'll trigger an alert. Pasting screenshot below 


<img width="1526" alt="image" src="https://user-images.githubusercontent.com/3794761/101073891-1a5e6280-356e-11eb-90cc-7c9d35c3f9e8.png">


<img width="1519" alt="image" src="https://user-images.githubusercontent.com/3794761/101074458-bf793b00-356e-11eb-9648-739cad9014e0.png">

kingthorin commented:
Will try to look this evening.

thc202 commented:
One of problems is that the proxy created by the scan rule does not have the Root CA cert thus failing to access the jQuery scripts. That can be addressed by adding to the test `setup` method `CachedSslCertifificateServiceImpl.getService().initializeRootCA(SslCertificateUtils.createRootCA());` but I'd suggest adding the required resources locally to not rely on the remote ones.

mrod23 commented:
Winded up adding the JQuery scripts directly to the CancelButton.html page. 

This is now passing the build. 

Ready for review, please

mrod23 commented:
@psiinon  / @thc202 
Can you please take a look at this MR when you get a chance? Think it's ready for approval.

psiinon commented on addOns/domxss/src/main/java/org/zaproxy/zap/extension/domxss/DomXssScanRule.java:
```diff
@@ -476,10 +477,14 @@ private void scanHelper(String attackVector, String url) throws DomAlertExceptio
+
             } catch (UnhandledAlertException uae) {
-                Stats.incCounter("domxss.vulns.input2");
+                Stats.incCounter("domxss.vulns.possibleXSSTriggers2");
```
I agree these names are more descriptive, but what if people are already using the existing ones?
No idea if thats the case, but if so then it would break things for them...

mrod23 replied:
Updated to possibleDomXSSTriggers2

psiinon commented:
Woah, the CancelButton.html file is huge :P
Including a copy of jquery makes it harder to maintain.
I'd much rather a simple html file with minimal vanilla JS as per the other tests. The vast majority of the JS in that file is not needed...

psiinon commented:
Other than those comments - its getting very close :)

thc202 commented:
Would you mind squash the commits?

psiinon commented:
As an aside, clicking on one button could change the state in a way that impacts another button.
For example, there could be a form with a 'Clear' button which just cleas the fields.
The rule could hit the clear button, clearing the fields, before hittin another button which would trigger the XSS.
Dont think that needs to be addressed in this PR as it could get v complicated, just worth thinking about for the future. If we have a simple native JS example which the rules works with the we would hopefully be able to tweak that to create a false negative example that we'd want to fix at some point...

kingthorin commented:
https://github.com/zaproxy/zap-extensions/pull/2638/checks?check_run_id=1524550292

To address the DCO requirement you'll need to sign-off the commit(s):

- https://github.com/zaproxy/zaproxy/blob/develop/CONTRIBUTING.md#developer-certificate-of-origin
- https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

mrod23 commented:
@psiinon  updated, pulled out the JQuery and using my own JavaScript. 
@thc202  squashed commits

mrod23 commented:
@psiinon  / @thc202 
I think this PR is ready to go now. Please review/approve when you get a chance

thc202 commented:
Thanks!

kingthorin reviewed (approved):
Thanks
